arwen/elrondapi/elrondei.go

```diff
@@ -796,14 +796,105 @@ func v1_3_transferValue(context unsafe.Pointer, destOffset int32, valueOffset in
 		return 1
 	}
 
-	err = output.Transfer(dest, send, 0, 0, big.NewInt(0).SetBytes(valueBytes), data, vmcommon.DirectCall)
+	err = output.Transfer(dest, sender, 0, 0, big.NewInt(0).SetBytes(valueBytes), data, vmcommon.DirectCall)
 	if arwen.WithFault(err, context, runtime.ElrondAPIErrorShouldFailExecution()) {
 		return 1
 	}
 
 	return 0
 }
 
+func extractIndirectContractCallArgumentsWithValue(
+	host arwen.VMHost,
+	destOffset int32,
+	valueOffset int32,
+	functionOffset int32,
+	functionLength int32,
+	numArguments int32,
+	argumentsLengthOffset int32,
+	dataOffset int32,
+) (dest []byte, value *big.Int, function []byte, args [][]byte, actualLen int32, err error) {
+	return extractIndirectContractCallArguments(
+		host,
+		destOffset,
+		valueOffset,
+		true,
+		functionOffset,
+		functionLength,
+		numArguments,
+		argumentsLengthOffset,
+		dataOffset,
+	)
+}
+
+func extractIndirectContractCallArgumentsWithoutValue(
+	host arwen.VMHost,
+	destOffset int32,
+	functionOffset int32,
+	functionLength int32,
+	numArguments int32,
+	argumentsLengthOffset int32,
+	dataOffset int32,
+) (dest []byte, value *big.Int, function []byte, args [][]byte, actualLen int32, err error) {
+	return extractIndirectContractCallArguments(
+		host,
+		destOffset,
+		0,
+		false,
+		functionOffset,
+		functionLength,
+		numArguments,
+		argumentsLengthOffset,
+		dataOffset,
+	)
+}
+
+func extractIndirectContractCallArguments(
+	host arwen.VMHost,
+	destOffset int32,
+	valueOffset int32,
+	hasValueOffset bool,
+	functionOffset int32,
+	functionLength int32,
+	numArguments int32,
+	argumentsLengthOffset int32,
+	dataOffset int32,
+) (dest []byte, value *big.Int, function []byte, args [][]byte, actualLen int32, err error) {
```

Comment:
return a struct instead of these multiple returns. small internal struct in this file would be enough.